### PR TITLE
MSC3005: Streaming Federation Events

### PR DESCRIPTION
**Pull request by ShadowJonathan**
[Rendered](https://github.com/ShadowJonathan/matrix-doc/blob/ws/proposals/3005-federation-streaming.md)

`Signed-off-by: Jonathan de Jong <jonathan@automatia.nl>`

### CODE REVIEW

File: proposals/3005-federation-websockets.md [hunk 1]
```diff
@@ -0,0 +1,142 @@
...
+encapsulating queued PDUs and EDUs and sending them through individual HTTP requests.
+
+This system can be slow or otherwise resource-intensive in a number of ways, through how each
+transaction request is largely the same (except content and maybe `txn` ), and how each goes through
+the entire HTTP request chain for every single transaction, which can hamper Time-To-Deliver over federation.
+
+## Proposal
+
+The proposal thus is to then introduce a new optionally-implementable WebSocket endpoint,
+`/federation/v2/stream` , which is then upgradable to a WebSocket connection.
```
Comment (ShadowJonathan) [inline proposals/3005-federation-websockets.md:14]: No idea if this should be `v2`

File: proposals/3005-federation-websockets.md [hunk 2]
```diff
@@ -0,0 +1,142 @@
...
+
+Servers over federation can implement this endpoint, if servers encounter 404 when querying this
+endpoint, it should be assumed the server does not implement WebSocket streaming, and instead
+`/federation/v1/send/{txn}` -based federation event-sending must be used.
+
+A connecting server must include `X-Matrix-Origin-Server` in their request to the stream endpoint.
```
Comment (ShadowJonathan) [inline proposals/3005-federation-websockets.md:20]: `X-Matrix-Origin-Server` is a fall-in for the `origin` key on `v1/send`, I initially thought this'd be "this" for the whole stream, and maybe some PDUs or EDUs being able to be defined to be "overridden" by an differently-sent server.

Though that may have security problems.

File: proposals/3005-federation-websockets.md [hunk 3]
```diff
@@ -0,0 +1,146 @@
...
+
+*This section outlines 2 different framing options I had in mind, please comment with thoughts.*
+
+*Anything included in square brackets is a proposal, not fully thought-out*
+
+#### Leading-type-byte frame
+
+In the WebSocket stream, a frame must be interpreted in the following fashion;
+
+[The leading byte denotes what type and format the data following it will be.]
```
Comment (ShadowJonathan) [inline proposals/3005-federation-websockets.md:32]: This approach is based around speed, a single byte (255 possible meanings) can be parsed quickly, however, the alternate approach below, can have infinite meanings at the cost of slower (null-terminated) parsing, and potentially more readability.

File: proposals/3005-federation-websockets.md [hunk 4]
```diff
@@ -0,0 +1,146 @@
...
+```
+
+`11` *or* `perr` returns an error for a PDU, it's payload is defined as follows;
+
+``` json
+[
+  "<PDU ID>",
+  "<Error in string format>"
+]
+```
```
Comment (ShadowJonathan) [inline proposals/3005-federation-websockets.md:75]: An array/tuple could potentially be quicker to deserialise, rather than an object with two (pretty much redundant) keys.

File: proposals/3005-federation-websockets.md [hunk 5]
```diff
@@ -0,0 +1,146 @@
...
+
+### "object"
+
+This proposal does not lock down the payload formatting down to JSON, it does make it a default.
+
+The recommendation is to use `?format=cbor` or the like to define it's payload object format, in the
+future.
```
Comment (ShadowJonathan) [inline proposals/3005-federation-websockets.md:82]: It's my intention to make this MSC introduce JSON-based streaming first, and have it be able to flawlessly adopt CBOR or other buffers in the future. (see comments above about sending a string on `10`/`pack`)

File: proposals/3005-federation-websockets.md [hunk 6]
```diff
@@ -0,0 +1,146 @@
...
+### Control Events
+
+Control events are defined as follows;
+
+``` json
+{
+  "op": 0, // specific control opcode
+  "data": null // arbitrary data for the opcode
+}
+```
```
Comment (ShadowJonathan) [inline proposals/3005-federation-websockets.md:93]: Opcode inside an opcode-defined stream?

Yeah, i know, this can be refactored into being just an "extension" of the opcodes, e.g. `F~` being the bit-prefix to a control event (which'd leave 16 control events ready to be defined).

So this can either be an opcode object, or the control-event opcodes can become `ctrl-` or `F~`-prefixed events.

File: proposals/3005-federation-websockets.md [hunk 7]
```diff
@@ -0,0 +1,146 @@
...
+
+| Opcode | Description               | Content                 | Comment                                                                      |
+| ------ | ------------------------- | ----------------------- | ---------------------------------------------------------------------------- |
+| 0      | Heartbeat                 | TS of origin server     | Can be sent by either server                                                 |
+| 1      | Heartbeat response        | TS of responding server |                                                                              |
+| 2      | Graceful shutdown request |                         | Useful for terminating the connection correctly and cleanly by either server |
+
+*The current proposal is unidirectional, it is possible to make this bidirectional.*
```
Comment (ShadowJonathan) [inline proposals/3005-federation-websockets.md:101]: This is a pretty big thing, I can make this bidirectional (which eliminates another connection to send events to the other server), but that'd change the architectural perspective of this.

Matrix, as of right now, has federation "senders" and "receivers", this'd make an extra option; federation "workers", which'd send and receive federation events simultaneously.

While more efficient on resources, it'll need to be scrutinised and considered closely.

File: proposals/3005-federation-websockets.md [hunk 8]
```diff
@@ -0,0 +1,146 @@
...
+
+*or*
+
+It's payload is defined as follows;
+
+``` json
+{
+  "id": "<PDU ID>"
+}
+```
```
Comment (ShadowJonathan) [inline proposals/3005-federation-websockets.md:66]: It being either a string or an object depends on the "object" capabilities, and if it's smarter to cause on the side of error and make everything an object, or just make everything "only as much as it needs to be", this stream event notes a PDU ack, and all the receiving server then needs is to know *what* is acknowledged, a string of the event ID, in this case.

**Review by ShadowJonathan — COMMENTED** [review]
Wrote some ~~sleep-deprived~~ comments that i know are absolute issues that need to be addressed here, this MSC requires more comments.

**Comment by ShadowJonathan** [comment]
Small PSA: I *might* refactor this MSC into a "generic streaming" spec instead, due to some insightful discussion i had.

File: proposals/3005-federation-streaming.md [hunk 9]
```diff
@@ -0,0 +1,412 @@
...
+
+Like the `Hello` frame, it is also encoded in UTF-8 `JSON`, and only after this frame has been
+received by the requesting side, should frame payload formatting be considered altered or freely
+defined.
+
+|    Field     | Description                        | Type                     | Omit default        |
+| :----------: | ---------------------------------- | ------------------------ | ------------------- |
+|    `fmt`     | Chosen format                      | `integer | string`       | (Cannot be omitted) |
+|    `ext`     | Echoed extension settings          | `object{string: value}`  | `{}`                |
+| `ext_op_map` | Mapped extension-frames-to-opcodes | `object{string: string}` | `{}`                |
```
Comment (ShadowJonathan) [inline proposals/3005-federation-streaming.md:125]: This can also be `ext_ops`, though i wasn't sure.

File: proposals/3005-federation-streaming.md [hunk 10]
```diff
@@ -0,0 +1,412 @@
...
+Acknowledges the PDU sent from the other side with an error.
+
+The payload must be an `array`, sized with 2 `value`s, defining each as follows;
+
+First value; a `string` responding to the specific PDU event being responded to.
+
+Second value; a free `value` detailing the error of processing this EDU. It should probably be a
+`string`, or an `object` with an `"error"` key, but this is not absolutely required.
+
+How server implementations react to PDU errors is out of scope for this proposal.
```
Comment (ShadowJonathan) [inline proposals/3005-federation-streaming.md:276]: The original spec is incredibly lax with how the "processing results" are defined, so I don't have a choice to make it just as ambiguous.

File: proposals/3005-federation-streaming.md [hunk 11]
```diff
@@ -0,0 +1,412 @@
...
+The first byte denotes the [opcode](#opcodes), effectively a `u8`/`uint8`.
+
+The next 4 bytes denote the length of the remaining payload, effectively `u32`/`uint32`.
+
+The remaining payload is defined per opcode, with formatting defined per-connection (see
+[`hello`](#hello) below).
+
+### Handshake
+
+Upon establishing a federation websocket connection, the requesting side needs to first send a
```
Comment (Cadair) [inline proposals/3005-federation-streaming.md:43]:
```suggestion
Upon establishing a federation streaming connection, the requesting side needs to first send a
```

Comment (ShadowJonathan) [reply proposals/3005-federation-streaming.md:43]: oh whoops

**Comment by ShadowJonathan** [comment]
Concerns from @Half-Shot re: this, that the spec should avoid bloat, this is a new protocol which'd be implemented in the spec, and matrix has so far been wanting to avoid reinventing the wheel, generally being in favour of using something else that's "out there" before using this protocol.

File: proposals/3005-federation-streaming.md [hunk 12]
```diff
@@ -0,0 +1,433 @@
...
+
+WebTransport *can* be complementary to future proposals detailing how exactly Federation Streams must
+or could be established.
+
+**HTTP/2**: A new HTTP paradigm, utilizing one connection, multiplexing streams that handle HTTP
+requests with compressed headers in binary.
+
+HTTP2 offers many optimization strategies, and has some interesting potential applications in
+parallelizing federation transactions, but eventually, the best optimization is simply to eliminate
+the HTTP request router-matching stack through which all those requests must eventually go through.
```
Comment (deepbluev7) [inline proposals/3005-federation-streaming.md:411]: Your proposed mechanism does not define, which port it will run on or how the port is discovered. It also does not define, how you would run multiple hosts on the same ip:port. (Well, it kinda does a bit in the frontloading section, but I don't think it is extensive enough.) It also doesn't really have a versioning field. All of the above is done using the http routing stack currently (http version field, hostnames, proxying, the v1 in `v1/send`). It is common to host multiple Matrix servers behind the same proxy, as such this should still be possible. I don't really see, how this proposal still allows it without introducing the same overhead.

On the other hand, this proposal does not include pipelining, multiplexing, stream priorities, which are all useful features to have and which you get using http/2 for "free". Those features make it possible to just have one persistent connection between servers and send all the other federation requests through it as well, while still prioritizing requests, that will block transaction processing (like prev events or state requests).

Your proposal actually looks very similar to http/2 or http/3 in its basic layout, but it removes the routing, which is a feature used for load balancing, fail over etc and in general is less powerful, less optimized and less supported. I don't really see any major benefits of your proposal over just optimizing the current protocol around parallel transactions and the servers around http/2+. The promised performance gains from the protocol itself sound minuscule, the big gains are probably in sending PDUs and EDUs outside of bigger transactions. This should reduce latency, but may increase CPU load by a lot (presence is known to use a lot of CPU, because it causes a lot of small transactions. Bundling them up in bigger transactions helps a lot).

So to summarize, HTTP/2 looks more attractive to me.

Comment (ShadowJonathan) [reply proposals/3005-federation-streaming.md:411]: ip:port is irrelevant to this proposal, this proposal defines an abstract streaming spec protocol, with specific implementation proposals  (TCP/WS) taking it to reach and carry this stream instead.

Versioning is an interesting problem i didn't realise before, i'll think about it. For now my assumption was to have all opcodes be defined once and then never updated, but this doesn't scale, obviously.

multiplexing is not in the scope of this streaming proposal, it is an application protocol, and multiplexing can happen underneath it. ergo with stream priorities, all this stream is supposed to do is send data, with the case of PDUs and EDUs these can be asynchronously processed, but ordering is up to either side, and every frame is as itself (expected to be less than 1kb in most cases, way too low of an amount of data to be relevant for prioritisation, maybe if bulk data like content is included in this, this'd becomes relevant).

> while still prioritizing requests, that will block transaction processing (like prev events or state requests).

This protocol doesn't say (and i should add that it *explicitly* doesn't) that every frame *must* to be responded to synchronously, it only says this about control frames (which even then is "should", to make it a priority on both sides, not an absolute one), so on that premise, these things simply arent needed.

> but it removes the routing, which is a feature used for load balancing, fail over etc and in general is less powerful, less optimized and less supported.

Can you explain this to me? Routing in HTTP is done on a per-request basis, HTTP paradigms meet every request with a response, HTTP2 also goes a step beyond that by adding pushing resources to clients to "step ahead" of that problem. Failing over a connection like this would be incredibly difficult, as it is effectively designed to be a single state machine on a single connection (what PDUs have been acked, what EDUs have, heartbeat nonces, etc.), so if that one fails, then effectively the recovery can only fail, and this protocol is designed to handle that (either with a "go away" frame voluntarily, or a dropped connection, and then having PDUs not responded to with acks)

> presence is known to use a lot of CPU, because it causes a lot of small transactions. Bundling them up in bigger transactions helps a lot

That is out of scope of the current proposal, and not something to be concerned about. It also doesn't make sense, how would it help them to bundle them in bigger transactions? On the basis of processing per persistence, the absolute processing time shouldn't matter, the only thing that increases load is the increased pressure on federation to constantly send these EDUs out to other servers (which is the overhead from HTTP routing internally).

**Comment by ShadowJonathan** [comment]
Some extra discussion on `#matrix-spec:matrix.org` has yielded the following feedback i'll be implementing;
- Versioning (Still thinking about it, I need to think of a proper way that doesn't inhabit the entire front-end of the stream session)
- Changing the assumption of the underlying connection to be a framed one.
  - This is done to deduplicate efforts of framing and make it possible to take advantage of underlying framing in many situations.
  - Pure-streaming connection types will have to implement a "shim" of this, e.g. plastering opcode+payload length in front to segment each chunk into a frame, this is up to those individual proposals to implement and specify (if needed)
  - > "the underlying connection MUST be framed, either through native capabilities (WebSocket frames), or emulated (TCP with payload length header)"
- Make the heartbeat opcodes completely optional, and put these into a dedicated extensions (with their extension frames being freely mappable)

**Comment by ShadowJonathan** [comment]
I believe that this MSC does no longer have much going for it at this stage.

While I believe that a binary&streaming-based solution can speed up raw transfers and decrease raw latency of federation traffic, it will increase implementation complexity, and possibly heighten the barrier to entry for a matrix homeserver.

I do ask the SCT to consider ways to speed up raw transfers, but I don't think I have confidence that this MSC, at this time, and possibly with me as the author, can achieve something like that.

Thus, thanks for your help with this, I am abandoning this MSC.

For anyone reading this at any point in the future, I invite them to take note from the proposal contents to make a better solution with a better context or perspective on this issue.
